src/dense_layout.rs

```diff
@@ -0,0 +1,213 @@
+        IndexSet::with_hasher(RandomState::default());
+    let mut bfs_order = Vec::with_capacity(n);
+    let mut seen: HashSet<usize> = HashSet::with_capacity(n);
+    next_level.insert(start);
+    while !next_level.is_empty() {
+        let this_level = next_level;
+        next_level = IndexSet::with_hasher(RandomState::default());
+        let mut found: Vec<usize> = Vec::new();
+        for v in this_level {
+            if !seen.contains(&v) {
+                seen.insert(v);
+                found.push(v);
+                bfs_order.push(v);
+            }
+        }
+        if seen.len() == n {
+            return bfs_order;
+        }
+        for node in found {
+            for (idx, v) in adj_matrix.index_axis(Axis(0), node).iter().enumerate() {
+                if *v != 0. {
+                    next_level.insert(idx);
+                }
+            }
+            for (idx, v) in adj_matrix.index_axis(Axis(1), node).iter().enumerate() {
+                if *v != 0. {
+                    next_level.insert(idx);
+                }
+            }
+        }
+    }
+    bfs_order
+}
+
+/// Find the best subset in the coupling graph
+///
+/// This function will find the best densely connected subgraph in the
+/// coupling graph to run the circuit on. It factors in measurement error and
+/// cx error if specified.
+///
+/// Args:
+///
+///     num_qubits (int): The number of circuit qubits
+///     coupling_adjacency (numpy.ndarray): An adjacency matrix for the
+///         coupling graph.
+///     num_meas (int): The number of measurement operations in the circuit
+///     num_cx (int): The number of CXGates that are in the circuit
+///     use_error (bool): Set to True to use the error
+///     symmetric_coupling_map (bool): Is the coupling graph symmetric
+///     error_matrix (numpy.ndarray): A 2D array that represents the error
+///         rates on the target device. The diagonal is the measurement error
+///         error rate on that qubit and then the other positions are the
+///         2q/cx error rate between the 2 indices.
+///
+/// Returns:
+///     (rows, cols, best_map): A tuple of the rows, columns and the best
+///     mapping found by the function. This can be used to efficiently create
+///     a sparse matrix that maps the layout of virtual qubits
+///     (0 to ``num_qubits``) to the physical qubits on the coupling graph.
+#[pyfunction]
+pub fn best_subset(
+    py: Python,
+    num_qubits: usize,
+    coupling_adjacency: PyReadonlyArray2<f64>,
+    num_meas: usize,
+    num_cx: usize,
+    use_error: bool,
+    symmetric_coupling_map: bool,
+    error_matrix: PyReadonlyArray2<f64>,
+) -> PyResult<(PyObject, PyObject, PyObject)> {
+    let coupling_adj_mat = coupling_adjacency.as_array();
+    let coupling_shape = coupling_adj_mat.shape();
+    let err = error_matrix.as_array();
+    let avg_meas_err = err.diag().mean().unwrap();
+    let best_result = (0..coupling_shape[0])
+        .into_par_iter()
```

Comment:
I just remembered that we should probably switch here like we do in stochastic swap based on the env variables. I extract that in #7702 into a shared function. So we probably should wait for that PR to merge and I can use the `eval_parallel_env()` function added there in this PR to switch between parallel and single threaded versions here.

Reply:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7740/commits/90860dc3a3bf3586a1851e409ded943a92d3409f